pkg/ee/whitelisted-registry-controller/reconcile.go

```diff
@@ -0,0 +1,229 @@
+      MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+      IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+      CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+      TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+      SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+   END OF TERMS AND CONDITIONS
+*/
+
+package whitelistedregistrycontroller
+
+import (
+	"context"
+	"fmt"
+
+	constrainttemplatev1beta1 "github.com/open-policy-agent/frameworks/constraint/pkg/apis/templates/v1beta1"
+	"go.uber.org/zap"
+
+	kubermaticapiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	WhitelistedRegistryCTName = "whitelistedregistry"
+	WhitelistedRegistryField  = "whitelisted_registry"
+)
+
+type Reconciler struct {
+	log          *zap.SugaredLogger
+	recorder     record.EventRecorder
+	masterClient ctrlruntimeclient.Client
+	namespace    string
+}
+
+func NewReconciler(log *zap.SugaredLogger, recorder record.EventRecorder, masterClient ctrlruntimeclient.Client, namespace string) *Reconciler {
+	return &Reconciler{
+		log:          log,
+		recorder:     recorder,
+		masterClient: masterClient,
+		namespace:    namespace,
+	}
+}
+
+// Reconcile reconciles the whitelisted registry
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Reconciling")
+
+	whitelistedRegistry := &kubermaticv1.WhitelistedRegistry{}
+	if err := r.masterClient.Get(ctx, request.NamespacedName, whitelistedRegistry); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debug("whitelisted registry not found, returning")
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, fmt.Errorf("failed to get whitelisted registry %s: %v", whitelistedRegistry.Name, err)
+	}
+
+	err := r.reconcile(ctx, whitelistedRegistry)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Eventf(whitelistedRegistry, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	return reconcile.Result{}, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, whitelistedRegistry *kubermaticv1.WhitelistedRegistry) error {
+	if whitelistedRegistry.DeletionTimestamp != nil {
+		if !kuberneteshelper.HasFinalizer(whitelistedRegistry, kubermaticapiv1.WhitelistedRegistryCleanupFinalizer) {
+			return nil
+		}
+
+		// Ensure Constraint with registry data
```

Comment:
Do we also want to delete the constraint template for this WhitelistedRegistry if it is deleted?

Reply:
good question. 

For now i decided to leave it as by itself(without an active Constraint), the ConstraintTemplate doesnt do anything, but we could consider removing it as well. WDYT?

Reply:
Both look good to me, I was asking because I noticed this is created in the reconcile loop but we didn't remove it when WhiteListedRegistery is removed, so I was wondering if we missed anything in the controller logic :)
but if the ContraintTemplate doesn't do anything, I am fine with leaving it